### PR TITLE
Reduce dashboard bootstrap payload

### PR DESCRIPTION
**Pull request by etr2460**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Currently, when loading a dashboard we send down the complete data payload for all datasources used within the dashboard. This is somewhat of a waste, because in the dashboard view the entire datasource isn't needed, only the pieces required to render the slices within the dashboard (as opposed to the explore view, where we need the entire datasource). When datasources have hundreds or thousands of columns or metrics, this can get out of hand quite quickly; we have a datasource that ends up being about 6MB of uncompressed JSON.

This PR adds a new function to the datasource model that filters the full data to only pieces required for a set of slices on a dashboard. This consists of:
- Removing unneeded columns
- Removing unneeded metrics
- Removing unneeded columns and metrics from the verbose_map
- Removing the datasource description

When determining what columns and metrics were required to render charts properly, I referenced the form_data definition from CONTRIBUTING.md to ensure that we included all the required columns and metrics for all the various charts.

### TEST PLAN
New unit tests. Also load the world bank dashboard before and after the change. See it render the same way, but require significantly less data

Before:
Screen Shot 2020-03-11 at 11 44 56 AM
Screen Shot 2020-03-11 at 11 45 10 AM

After:
Screen Shot 2020-03-11 at 11 43 43 AM
Screen Shot 2020-03-11 at 11 44 10 AM

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
to: @john-bodley @villebro @mistercrunch @kristw

### CODE REVIEW

**Comment by kristw** [comment]
I love this and fully support the idea, but will leave ppl with more python expertise to approve.

File: superset/connectors/base/models.py
Comment (villebro) [inline superset/connectors/base/models.py]: Two that I believe are missing here: `series` and `entity`.

File: superset/connectors/base/models.py
Comment (villebro) [inline superset/connectors/base/models.py]: I might be wrong here, but isn't this function already implemented in `from superset.utils.core import get_metric_name()`?

**Review by villebro — COMMENTED** [review]
First pass comments. This is definitely a move in a good direction. I anticipate there will be regressions due to this, but they should be easy to fix, and in the grand scheme of things definitely worth doing.

File: superset/connectors/base/models.py
Comment (john-bodley) [reply superset/connectors/base/models.py]: @etr2460 [this](https://github.com/apache/incubator-superset/blob/master/CONTRIBUTING.md#chart-parameters) could be a good resource for the relevant params.

File: superset/views/core.py
Comment (john-bodley) [inline superset/views/core.py]: Could we use a `defaultdict(lambda: defaultdict(list))` ? This means the logic in lines 1795 - 1801 could be simplified.

File: superset/connectors/base/models.py
Comment (etr2460) [reply superset/connectors/base/models.py]: note that `percent_metrics` is referenced in the function below because it's an array of metrics and not an individual one. It looks like I missed secondary_metric though. I'll refactor so that those are included in a constant as well.

File: superset/connectors/base/models.py
Comment (etr2460) [reply superset/connectors/base/models.py]: I don't actually think so, that function gets the "label" field off a metric instead of the "metric_name" one.  And I don't see the `label` param anywhere in the `BaseMetric` class, only the `metric_name` one

File: superset/connectors/base/models.py
Comment (etr2460) [reply superset/connectors/base/models.py]: added, good catch!

File: superset/views/core.py
Comment (etr2460) [reply superset/views/core.py]: i don't think this actually helps anything because we need to keep multiple fields in the dict of dicts, the datasource and a list of slices

**Review by etr2460 — COMMENTED** [review]
addressed comments

File: superset/connectors/base/models.py
Comment (etr2460) [reply superset/connectors/base/models.py]: I looked further, and it seems like the metric field is a string if it's a `metric_name`, but an object with a label param if it's an `adhoc_metric`. Although I'm not sure if these labels would actually show up in the datasource, I'll add logic to handle them

Comment (etr2460) [reply superset/connectors/base/models.py]: updated!

Comment (villebro) [reply superset/connectors/base/models.py]: The weird thing is what you're describing sounds more or less exactly like the other metrics that `get_metric_name` deals with. We don't need to deal with that now as it's outside the scope of this PR, but we definitely need to figure this out in the long run.

Comment (etr2460) [reply superset/connectors/base/models.py]: I think it's probably a confusion between "metrics" in form data and the "metric" model in Superset's backend. I think the "metrics" in form data are either `metric.metric_name` where `metric` is the metric model or an `AdHocMetric` in the JS frontend, which might not have any association with the metric model (i'm uncertain here). 

We could probably improve this by standardizing on a single typedef/naming scheme and then migrating all the chart form_data, but this isn't an easy thing to do (and will only get harder once charts are more pluggable and we can't guarantee what type of data might be in form_data)

File: superset/connectors/base/models.py
Comment (ktmud) [inline superset/connectors/base/models.py]: Maybe you can loop throw `form_data` keys and check whether it's in any of the consts instead.

```python
        metric_names = []
        column_names = []

        for slc in slices:
            # pull out all required columns from the form_data
            for param, value in slc.form_data.items():
                if param == 'adhoc_filters':
                    for filter_ in value:
                        if filter_["clause"] == "WHERE" and filter_.get("subject"):
                            column_names.add(filter_.get("subject"))
                elif param in METRICS_FORM_DATA_PARAMS:
                    metric_names += map(utils.get_metric_name, value)
                    # extra logic to handle adhoc metrics with column references
                    column_names += [
                        metric["column"].get("column_name")
                        for metric in value
                        if isinstance(metric, dict) and isinstance(metric["column"])
                    ]
                elif param in METRIC_FORM_DATA_PARAMS:
                    metric_names.append(utils.get_metric_name(value))
                elif param in COLUMNS_FORM_DATA_PARAMS:
                    column_names += value
                elif param in COLUMN_FORM_DATA_PARAMS:
                    column_names.append(value)

        metric_names = set(metric_names)
        column_names = set(column_names)
```

Btw the consts should probably be `set()`.

**Comment by ktmud** [comment]
Awesome work, @etr2460 ! Added some style nits.

File: superset/views/core.py
Comment (etr2460) [reply superset/views/core.py]: I don't believe the datasource can be the key as it's not hashable. I think your first solution will work though

File: superset/connectors/base/models.py
Comment (etr2460) [reply superset/connectors/base/models.py]: I think we need this just in case `form_data[param]` equals `''` or `0` or something else not iterable

Comment (etr2460) [reply superset/connectors/base/models.py]: although I think we can do `for metric in (form_data.get(param) or [])` and that should do the trick. Will modify and remove the if statement

File: superset/connectors/base/models.py
Comment (etr2460) [reply superset/connectors/base/models.py]: `form_data` can have a large number of parameters in it, I think it's a bit more efficient (if a couple extra lines of code) to only iterate through the ones we care about.

However, I went through the loop again and cleaned up some unnecessary checks

**Review by etr2460 — COMMENTED** [review]
@ktmud adressed comments

File: superset/views/core.py
Comment (ktmud) [reply superset/views/core.py]: You can probably create a hash function in the datasource class?

```
def ___hash__(self):
    return self.uid
```

Comment (etr2460) [reply superset/views/core.py]: i've added a `__hash__` and an `__eq__` function as according to https://docs.python.org/3/glossary.html#term-hashable

**Comment by etr2460** [comment]
@john-bodley @villebro this is ready for another review

File: superset/connectors/base/models.py
Comment (john-bodley) [inline superset/connectors/base/models.py]: Should there be similar logic for ad-hoc metrics here as well?

File: superset/connectors/base/models.py
Comment (john-bodley) [inline superset/connectors/base/models.py]: Should we use the [`is_adhoc_metric`](https://github.com/apache/incubator-superset/blob/7a91498cf1a9e56d4b3d7b3805076137525ea277/superset/utils/core.py#L944) method here to improve readability and adhering to the DRY principle?

File: superset/connectors/base/models.py
Comment (john-bodley) [inline superset/connectors/base/models.py]: Could we use `metric` or `m` rather than `o`?

File: superset/connectors/base/models.py
Comment (john-bodley) [inline superset/connectors/base/models.py]: Could we use `column` or `col` rather than `o`?

File: superset/views/core.py
Comment (john-bodley) [inline superset/views/core.py]: Nit. Remove the `.keys()` as this is implicit when iterating over a dictionary.

File: superset/views/core.py
Comment (john-bodley) [inline superset/views/core.py]: Could we make this a dictionary comprehension, 

```python
datasources_payload = {
    datasource.uid: datasource.data_for_slices(slices) for datasource, slices in datasources.items()
}
```

File: superset/connectors/base/models.py
Comment (villebro) [reply superset/connectors/base/models.py]: You're right, my bad!

File: superset/connectors/base/models.py
Comment (villebro) [reply superset/connectors/base/models.py]: I've been digesting this today, and I still think something isn't right here. If you look at `viz.py`, it's full of references to metrics in form_data. Copy/paste from `viz.py`: `utils.get_metric_names(self.form_data.get("metrics") or [])`. Are these really regular ad-hoc metrics that have been generated by Superset controls? I put a break point on `viz.py` with a viz with an ad-hoc metric, and that had the metric name in `label`, not `metric_name`. Is there a chance that we have different metric signature conventions in different parts of the codebase?

File: superset/connectors/base/models.py
Comment (etr2460) [reply superset/connectors/base/models.py]: oo good catch

File: superset/connectors/base/models.py
Comment (etr2460) [reply superset/connectors/base/models.py]: possibly. i'll add it

note that this has been pretty difficult to test, and even if a column is missing for a chart, I think the only implication is that we might lose custom formatting or a custom name for a column. It shouldn't break the entire chart

File: superset/connectors/base/models.py
Comment (etr2460) [reply superset/connectors/base/models.py]: yup, this was copied from above, but  agree a real name is better

File: superset/connectors/base/models.py
Comment (etr2460) [reply superset/connectors/base/models.py]: yup, this was copied from above, but  agree a real name is better

File: superset/views/core.py
Comment (etr2460) [reply superset/views/core.py]: someday i'll learn python

File: superset/views/core.py
Comment (etr2460) [reply superset/views/core.py]: done

**Review by etr2460 — COMMENTED** [review]
@john-bodley comments addressed

File: superset/connectors/base/models.py
Comment (john-bodley) [reply superset/connectors/base/models.py]: @etr2460 for an internal project to deal with scalar (metric) and list (metrics) I created a utility metric, 

```python
def get_iterable(x: Any) -> List:
   """
   Get an iterable (list) representation of the object.
   
   :param x: The object
   :returns: An iterable representation
   """
 
  return x if isinstance(x, list) else [x]
```

thus you could combine  the `METRICS_FORM_DATA_PARAMS` and `METRIC_FORM_DATA_PARAMS` lists (DRY) and do something of the form, 

```python
for param in ["metrics", "metric", ...]:
    for metric in get_iterable(form_data.get(param)):
        ...
```

**Comment by john-bodley** [comment]
@etr2460 it seems like you're tested this locally a bit and I know that @villebro has some questions/concerns but it would be good to potentially understand what type of regressions we may expect if say the current logic wasn't watertight given the various nuances of how entities are defined at the visualization and form-data level.

**Comment by etr2460** [comment]
So I took a look to see what code on the frontend depends on various properties on the datasource. You can see all the places charts depend on the datasource here: https://github.com/apache-superset/superset-ui-plugins/search?p=1&q=datasource&unscoped_q=datasource

From what I can see, the datasource is used primarily for two reasons:
1. To add special number formatting to a metric as in: https://github.com/apache-superset/superset-ui-plugins/blob/17a5f64312326e708a81bab4a45578fd181b2e1e/packages/superset-ui-legacy-preset-chart-big-number/src/BigNumber/transformProps.js#L88-L92
2. To format the name of a metric differently as in: https://github.com/apache-superset/superset-ui-plugins/blob/be604e24fd9c7ac715e448673c56bed40c5344ff/packages/superset-ui-legacy-plugin-chart-partition/src/transformProps.js#L43

All of the instances i've tested haven't broken any charts entirely, so I think the main source of regressions would be formatting regressions (e.g. rendering 0.03 instead of 3% or "metric_name_not_pretty" instead of "Pretty Metric Name"). I think that we could merge this now and deploy internally, being ready to revert fast if needed or patch with any minor issues that show up. Alternatively I could put this behind a feature flag and only enable it in our deployment to start if we thought that was safer. Up to @john-bodley and @villebro (and anyone else who has input)

**Comment by villebro** [comment]
Like I said, this feature has obvious performance upside, but I expect some small regressions to come along with it. While they should be easy to fix, they could be annoying for anyone who prefers stability over performance. I really like the idea of putting this behind a feature flag for now, as it gives us the best of both worlds.

**Comment by etr2460** [comment]
@villebro feature flag added!

**Review by villebro — APPROVED** [review]
LGTM! I'll be activating this flag on my devbox too, I can pitch in if I surface any weird bugs.

**Comment by john-bodley** [comment]
@etr2460 could you address my comments before merging?

File: superset/connectors/base/models.py
Comment (etr2460) [reply superset/connectors/base/models.py]: done!

**Review by john-bodley — APPROVED** [review]
LGTM

**Pull request merged**
